Users count PR cleanup

mike-lvov opened the pull request:
This PR addresses cleanup around some of the changes made in https://github.com/sparkletown/sparkle/pull/1108

Or particular note are the review comments in: https://github.com/sparkletown/sparkle/pull/1108#pullrequestreview-580000514

0xdevalias commented on src/components/molecules/EventDisplay/EventDisplay.tsx:
```diff
@@ -16,21 +17,26 @@ interface EventDisplayProps {
-        isLiveEvent && "schedule-event-container_live"
-      }`}
-    >
+    <div key={event.name} className={containerClasses}>
```
This shouldn't need the `key` prop at all, as it's not within a loop.

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/components/molecules/EventDisplay/EventDisplay.tsx:
```diff
@@ -16,21 +17,26 @@ interface EventDisplayProps {
     event.start_utc_seconds + event.duration_minutes * 60 >
       getCurrentTimeInUTCSeconds();
 
+  const containerClasses = classNames("schedule-event-container", {
+    "schedule-event-container_live": isLiveEvent,
+  });
```
Is this `_live` meant to be a 'variant' (aka modifier)? If so, i've been loosely following BEM in my recent styling cleanups, which would make this `--live`:

- http://getbem.com/naming/

0xdevalias commented on src/components/molecules/LiveSchedule/LiveEvent.tsx:
```diff
@@ -1,64 +0,0 @@
-import React from "react";
```
What was this file refactored/renamed into, if anything?

mike-lvov replied:
It was the copy of `EventDisplay` component

0xdevalias replied:
Cool, thanks.

0xdevalias commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,23 +20,17 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
 import "./RoomModal.scss";
 
-interface RoomModalProps {
+type RoomModalProps = {
```
```suggestion
export type RoomModalProps = {
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -81,7 +75,6 @@ export const RoomModal: React.FC<RoomModalProps> = ({ onHide, room }) => {
               </div>
               <div className="col">
                 <RoomModalOngoingEvent
-                  roomTitle={room.title}
```
Guessing this was just never used/needed?

mike-lvov replied:
It was used in `tracking user's location` logic before. Have refactored that

0xdevalias replied:
Makes sense.

0xdevalias commented on src/hooks/useKeyboardControls.ts:
```diff
@@ -34,6 +35,15 @@ export const useKeyboardControls: ReactHook<UseKeyboardControlsProps, void> = ({
   const { profile } = useUser();
   const { row, column } = profile?.data?.[venueId] ?? {};
 
+  const { enterRoom } = useRoom(room);
```
Where is `room` coming from for this?

Not sure if we would necessarily want this hook within this function, do we already use it in the component that calls `useKeyboardControls` or similar? Could we maybe just pass down `enterRoom` through the props or similar?

mike-lvov replied:
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/hooks/useRoom.ts:
```diff
@@ -13,11 +14,13 @@ export const useRoom = (room: Room) => {
+    () => venues?.find((venue) => roomUrl.endsWith(`/${venue.id}`)),
+    [roomUrl, venues]
   );
   const roomId = roomVenue ? roomVenue.name : room.url;
```
Since this is used with `useRecentLocationUsers`, I still don't believe `room.url` is correct here. It should be the `venueName/roomName` slug.

mike-lvov replied:
It is not related to slug/url lastSeenIn's logic

0xdevalias replied:
I believe it may be?

```
const roomId = roomVenue ? roomVenue.name : room.url;
const { recentLocationUsers } = useRecentLocationUsers(roomId);
```

If we are trying to get `recentLocationUsers` with `room.url`, that isn't using the `venueName/roomName` slug, so will fail to retrieve them. Or so I would think? (I haven't looked deeply into the logic of `useRecentLocationUsers` to verify this yet)

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/pages/VenuePage/VenuePage.tsx:
```diff
@@ -93,57 +94,49 @@ const VenuePage: React.FC = () => {
-    if (!venueId || !user || !profile) return;
+    }
 
     updateProfileEnteredVenueIds(profile?.enteredVenueIds, user?.uid, venueId);
```
It seems we already have `userId` defined in this file, so can we narrow the scope of the checks/etc here to use that rather than the full user object?

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
-  userId: string,
-  newLocationData: LocationData = {}
-) => {
+type UpdateLocationDataProps = {
```
```suggestion
export type UpdateLocationDataProps = {
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
 };
 
-export interface TrackLocationProps {
+type SetNewLocationDataProps = {
```
```suggestion
export type SetNewLocationDataProps = {
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
 const LOCATION_INCREMENT_SECONDS = 10;
 const LOCATION_INCREMENT_MS = LOCATION_INCREMENT_SECONDS * 1000;
 
 type LocationData = Record<string, number>;
```
```suggestion
export type LocationData = Record<string, number>;
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
-import { openRoomUrl, openUrl, venueInsideUrl } from "./url";
+import { openRoomUrl } from "./url";
 
 const LOCATION_INCREMENT_SECONDS = 10;
 const LOCATION_INCREMENT_MS = LOCATION_INCREMENT_SECONDS * 1000;
```
Might make sense for these to be moved to `settings.ts`?

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
+type SetNewLocationDataProps = {
   userId: string;
   locationName: string;
   lastSeenIn?: LocationData;
```
This doesn't seem to be used in the function anymore?

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
+type SetNewLocationDataProps = {
   userId: string;
   locationName: string;
   lastSeenIn?: LocationData;
-}
+};
 
-export const trackLocationEntered = ({
+export const setNewLocationData = ({
```
The naming of this feels weird.. almost reversed. I would expect that `setNewLocationData` would override everything, and `updateLocationData` would probably do the more selective update.

Would probably rename `setNewLocationData` to just `setLocationData` to match the same 'format' of `updateLocationData` as well.

(Doesn't necessarily have to follow the above, but the distinction/usages of them feels a bit confusing at the moment, so if we can improve that, that would be my aim here)

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
 };
 
-export const enterVenue = (venueId: string) => openUrl(venueInsideUrl(venueId));
+type UpdateCurrentLocationDataProps = {
```
```suggestion
export type UpdateCurrentLocationDataProps = {
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
+  userId,
+  profileLocationData,
+}: UpdateCurrentLocationDataProps) => {
+  const [locationName] = Object.keys(profileLocationData);
```
This is just going to set `locationName` to be the first array item returned from `Object.keys` yeah? That seems, wrong? Maybe?

What's the intended effect here?

mike-lvov replied:
The intended effect is to update the current location, without rewriting it. profileLocationData can only have 1 key at any point of time

0xdevalias replied:
Right, that makes sense. Perhaps adding a comment saying something like that would be useful, as it wasn't immediately obvious to me.

0xdevalias replied:
See also https://github.com/sparkletown/sparkle/pull/1129#discussion_r569931904

0xdevalias replied:
@mike-lvov can you please add a comment to the code here explaining what the intent/reasoning of this line is, so that future developers understand why we are only updating a single key here, and the assumptions/etc this was written with.

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
+  });
+};
+
+type ClearLocationData = {
```
```suggestion
export type ClearLocationDataProps = {
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
+  userId: string;
+};
+
+export const clearLocationData = ({ userId }: ClearLocationData) => {
```
Since we only have a single parameter, it's probably fine to just make it `userId: string` directly, and not wrap it in object destructuring. Which also means we can get rid of the `ClearLocationData` type above.

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,52 +1,83 @@
-  trackLocationEntered({
+  setNewLocationData({
     userId,
     locationName: room.url,
```
This shouldn't be using `room.url`, rather it should be using `venueName/roomName` slug.

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -55,26 +86,32 @@ export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
 
 export const useUpdateTimespentPeriodically = (
-  user: UserInfo | undefined,
-  roomName: string
+  roomName: string,
+  userId?: string
 ) => {
```
Why was the position of these parameters changed? Guessing because TypeScript wanted the optional param to be last? We should probably change this to use a 'props object' with object destructuring + named keys instead of positional parameters.

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -55,26 +86,32 @@ export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
     () => {
-      // Time spent is currently counted multiple time if multiple tabs are open
-      if (!user || !roomName) return;
+      // @debt time spent is currently counted multiple time if multiple tabs are open
```
[nit]
```suggestion
      // @debt time spent is currently counted multiple times if multiple tabs are open
```

mike-lvov replied:
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/509bf9e36436cddbbfb4c24212ca2e48f29b2a6c

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -55,26 +86,32 @@ export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
-      const increment = firebase.firestore.FieldValue.increment(
-        LOCATION_INCREMENT_SECONDS
-      );
+      const doc = `users/${userId}/visits/${roomName}`;
```
I still feel like we would be better to use the more verbose method of `firestore.doc(doc).collection('users').doc(userId)...etc` instead of this path, as I believe as it currently is, this would potentially allow a SQL injection/path traversal sort of issue.

mike-lvov replied:
fixed in
https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/components/templates/PartyMap/components/Map/Map.tsx:
```diff
@@ -84,11 +84,13 @@ export const Map: React.FC<MapProps> = ({
         venueId,
         userUid,
       })(row, column);
-      trackLocationEntered({ userId: userUid, locationName: venueName });
+
+      setNewLocationData({ userId: userUid, locationName: venueName });
```
I kind of feel like the old naming of this (`trackLocationEntered`) was better.

mike-lvov replied:
This naming aligns with the other function names `updateCurrentLocation`, `clearLocation`

0xdevalias replied:
Makes sense

0xdevalias reviewed (commented):
Some early review notes to align before the final review.

mike-lvov commented on src/components/molecules/LiveSchedule/LiveSchedule.tsx:
```diff
@@ -42,12 +42,8 @@ const LiveSchedule: FC = () => {
-        event={event}
-      />
+    return events.map((event) => (
+      <EventDisplay venue={relatedVenueFor(event)} event={event} />
```
> If we're calculating the venue from the event anyway (with relatedVenueFor(event)), is it worth having it as a separate prop?

It is not the only place, where `EventDisplay is used`. Not every time you calculate venue, based on the event.

0xdevalias replied:
Cool, that makes sense. Fine for it to stay as is then.

mike-lvov commented on src/components/molecules/NavBar/NavBar.scss:
```diff
@@ -487,7 +487,7 @@ $border-radius: 28px;
       background-color: rgba($black, 0.3);
     }
-    &.schedule-event-container_live {
+    &.schedule-event-container--live {
```
This is legacy style logic and I didn't want to mess with it

0xdevalias replied:
Makes sense

mike-lvov commented on src/components/molecules/LiveSchedule/LiveSchedule.scss:
```diff
@@ -45,7 +45,7 @@
       background-color: rgba($black, 0.3);
     }
-    &.schedule-event-container_live {
+    &.schedule-event-container--live {
```
This is legacy style logic and I didn't want to mess with it

0xdevalias replied:
Makes sense

mike-lvov commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,15 +21,40 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
-  const { enterRoom, recentRoomUsers } = useRoom(room);
+  const { enterRoom, recentRoomUsers } = useRoom(room, currentVenueName);
 
   useMousetrap({
```
I would still prefer having it here instead of <Map> `useKeyboardControls`.

1) We want to only have a listener when we need it.
2) It will require hacks to have `enterRoom` passed there since to have `enterRoom` we need to be sure there is selected room. (One way to do that would be to have noop fn returned from useRoom, but I don't want to go that way as it seems a bit hacky)

0xdevalias replied:
From memory, the purpose of it within the `useKeyboardControls` was so that when you are on the map, and you come across a room, you can press enter to 'bring it up' (or maybe to enter it, or similar.. I'm not 100%, as this was keeping the existing legacy logic, even though I refactored the 'how' of it)

My desire for this not to be here is to avoid us 'spreading' keyboard control logic across the app.

A question to ask here is whether we specifically need this keyboard control at all, be it in `useKeyboardControls`, or in `RoomModal`. It somewhat feels like it doesn't do a lot for us. If our desire is that we can enter the room from the room modal when this dialog is open, it may just also be possible by using existing buttons/html features. eg. 'submitting a form', or focussing the 'enter' button so that if you press enter, that's the first thing that will be 'actioned'.

If we decide that we can't use basic html features, and that we need to keep it in `RoomModal`, then as a 'final resort', we will need to change it so that its not using a global binding, but instead is 'scoped' to only register within the containing `div`. We can get that ref with `useRef`, but I believe there is also a bug in `useMousetrap` currently (detailed either in `@debt` comments, or maybe in a backlog item; I can't remember specifically) where I don't believe it will work properly with a 'non-global' binding. So there will be a bugfix required for that as well.

mike-lvov replied:
Let's keep the modal-keyboard thing convo here.

mike-lvov replied:
https://github.com/sparkletown/internal-sparkle-issues/issues/303

mike-lvov commented on src/components/molecules/NavSearchBar/NavSearchBar.tsx:
```diff
@@ -82,11 +83,9 @@ const NavSearchBar = () => {
     <div className="nav-search-links">
       <div className="nav-search-icon" />
       <NavSearchBarInput value={searchQuery} onChange={setSearchQuery} />
-
```
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

mike-lvov commented on src/components/templates/Audience/Audience.tsx:
```diff
@@ -142,6 +142,7 @@ export const Audience: React.FunctionComponent = () => {
   const venue = useSelector(currentVenueSelectorData);
   const { recentVenueUsers } = useRecentVenueUsers();
 
+  const userId = user?.uid;
```
fixed in https://github.com/sparkletown/sparkle/pull/1129/commits/53f57c40e24983470b7552d4ebed6f1715ff74ba

0xdevalias commented on src/hooks/useRoom.ts:
```diff
@@ -2,38 +2,42 @@ import { useCallback, useMemo } from "react";
-  const roomId = roomVenue ? roomVenue.name : room.url;
+
+  // @debt we should replace externalRoomSlug with preferrably room id
+  const roomId = roomVenue
+    ? roomVenue.name
+    : getExternalRoomSlug({ roomTitle: room.title, currentVenueName });
```
```suggestion
  const roomId = roomVenue
    ? roomVenue.name
    : getExternalRoomSlug({ venueName: currentVenueName, roomTitle: room.title });
```

mike-lvov replied:
fixed

0xdevalias commented on src/pages/VenuePage/VenuePage.tsx:
```diff
@@ -93,57 +94,49 @@ const VenuePage: React.FC = () => {
-
   useInterval(() => {
-    updateUserLocationToCurrentVenue();
+    if (!userId || !profile?.lastSeenIn) return;
+
+    updateCurrentLocationData({
+      userId,
+      profileLocationData: profile.lastSeenIn,
+    });
```
What will happen if they don't have `profile.lastSeenIn` (is that even possible?). Won't that make this never run? Or is this more about `profile` not being present? In which case, would we be better to just make it `!profile` instead?

Since `updateCurrentLocationData` is only taking the first entry from `profileLocationData` anyway, and then updating the time on that key, it feels weird that we need to pass the full thing through. Wouldn't it just be simpler to pass the name of the key we want to update through directly, and reduce all of this additional complexity?

Or alternatively, even though I believe you said that there can/should only ever be a single entry in `lastSeenIn`, since nothing actually enforces that currently, would it not be more robust to code `updateCurrentLocationData` to update every location key that is present? Or something to that effect?

mike-lvov replied:
if profile.lastSeenIn is empty, there is nothing to update, thus it will not update. We do set locations in a bunch of places. I don't know the way it is going to evolve, so I'm not sure it is efficient to think that ahead. If there is change in logic in the future, a changer should understand the way it works.

profileLocationData has only one key inside. I could pass this key, but why? What if we decide to change the logic in the future?

0xdevalias replied:
> profileLocationData has only one key inside. I could pass this key, but why? What if we decide to change the logic in the future?

My desire was to align what is passed into the function, with what the function is doing. I don't feel particularly strongly about changing the parameter, your reasoning seems to make sense RE: continuing to use `LocationData`

> I don't know the way it is going to evolve, so I'm not sure it is efficient to think that ahead. If there is change in logic in the future, a changer should understand the way it works.

Fair enough. I guess if you haven't already, can you add a comment or two in `updateCurrentLocationData` explaining why it's only taking the first param/updating that, not the entire collection/etc. That would satisfy my request here.

0xdevalias commented on src/pages/VenuePage/VenuePage.tsx:
```diff
@@ -93,57 +94,49 @@ const VenuePage: React.FC = () => {
-    updateProfileEnteredVenueIds(profile?.enteredVenueIds, user?.uid, venueId);
-  }, [profile, user, venueId]);
+    updateProfileEnteredVenueIds(profile?.enteredVenueIds, userId, venueId);
+  }, [profile, userId, venueId]);
```
It would be better for this function to not have to pass in `enteredVenueIds` directly, but instead to use firebase's `arrayUnion` helper

- https://firebase.google.com/docs/firestore/manage-data/add-data#update_elements_in_an_array

It would also be good to use a props object with named keys since there are multiple parameters.

mike-lvov replied:
[OOS] https://github.com/sparkletown/internal-sparkle-issues/issues/302

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
+export type LocationData = Record<string, number>;
 
-type LocationData = Record<string, number>;
+export type UpdateLocationDataProps = {
```
As per https://github.com/sparkletown/sparkle/pull/1062#issuecomment-772978074 this should use `interface` rather than `type`

mike-lvov replied:
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570165378

0xdevalias replied:
As per https://github.com/sparkletown/sparkle/pull/1129#discussion_r570166887 I believe this is still a valid request, and aligns with the other changes made in this PR to align 'props' objects to using `interface`.

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
     lastSeenIn: newLocationData,
   });
 };
 
-export interface TrackLocationProps {
+export type SetLocationDataProps = {
```
As per https://github.com/sparkletown/sparkle/pull/1062#issuecomment-772978074 this should stay as `interface`

mike-lvov replied:
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570165378

0xdevalias replied:
As per https://github.com/sparkletown/sparkle/pull/1129#discussion_r570166887 I believe this is still a valid request, and aligns with the other changes made in this PR to align 'props' objects to using `interface`.

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
+  });
+};
+
+export type UpdateCurrentLocationDataProps = {
```
As per https://github.com/sparkletown/sparkle/pull/1062#issuecomment-772978074 this should use `interface` rather than `type`

mike-lvov replied:
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570165378

0xdevalias replied:
As per https://github.com/sparkletown/sparkle/pull/1129#discussion_r570166887 I believe this is still a valid request, and aligns with the other changes made in this PR to align 'props' objects to using `interface`.

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
+};
 
-type EnterExternalRoomProps = {
+export type EnterExternalRoomProps = {
```
As per https://github.com/sparkletown/sparkle/pull/1062#issuecomment-772978074 this should use `interface` rather than `type`

mike-lvov replied:
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570165378

0xdevalias replied:
As per https://github.com/sparkletown/sparkle/pull/1129#discussion_r570166887 I believe this is still a valid request, and aligns with the other changes made in this PR to align 'props' objects to using `interface`.

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
+export type UpdateCurrentLocationDataProps = {
+  userId: string;
+  profileLocationData: LocationData;
+};
+
+export const updateCurrentLocationData = ({
+  userId,
+  profileLocationData,
+}: UpdateCurrentLocationDataProps) => {
+  const [locationName] = Object.keys(profileLocationData);
+
+  updateLocationData({
+    userId,
+    newLocationData: { [locationName]: getCurrentTimeInMilliseconds() },
   });
 };
 
-export const enterVenue = (venueId: string) => openUrl(venueInsideUrl(venueId));
+export const clearLocationData = (userId: string) => {
+  updateLocationData({ userId, newLocationData: {} });
+};
 
-type EnterExternalRoomProps = {
+export type EnterExternalRoomProps = {
   userId: string;
-  room: Room;
+  roomUrl: string;
+  roomId: string;
```
Since we need both `id` and `url` from `Room`, what makes it better to pass them as individual props than just passing `Room`? Are there places we need to use this where we won't have a `Room` object available?

mike-lvov replied:
roomId is the slug, not the room.id

mike-lvov replied:
changed `roomId` => `roomSlug`

0xdevalias replied:
[nit] Since this is just being passed into the `locationName` field, which we use elsewhere to refer to the 'location slug', should we also name this param similarly? (eg. `locationName` instead of `roomSlug`?)

0xdevalias replied:
fixed in 200c94f53638642cd3f70418f66b2accb8f392fe

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
+  openRoomUrl(roomUrl);
+};
+
+export type UseUpdateTimespentPeriodicallyProps = {
```
As per https://github.com/sparkletown/sparkle/pull/1062#issuecomment-772978074 this should use `interface` rather than `type`

mike-lvov replied:
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570165378

0xdevalias replied:
fixed in 200c94f53638642cd3f70418f66b2accb8f392fe

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
-  roomName: string
-) => {
-  const shouldUseInterval = user && roomName;
+// @dept I don't think this functinality works correctly, since we only log `internal venues` in this piece of code
```
```suggestion
// @debt I don't think this functionality works correctly, since we only log `internal venues` in this piece of code
```

mike-lvov replied:
fixed

0xdevalias replied:
It seems like it may not have been? (or if it was, the code may not have been pushed?)

Since we will likely be searching for this in the codebase with `@debt` (rather than `@dept`), would be good to have it aligned.

```suggestion
// @debt I don't think this functionality works correctly, since we only log `internal venues` in this piece of code
```

0xdevalias replied:
fixed in 200c94f53638642cd3f70418f66b2accb8f392fe

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
-) => {
-  const shouldUseInterval = user && roomName;
+// @dept I don't think this functinality works correctly, since we only log `internal venues` in this piece of code
+// Could be also beneficial to log external rooms' timespent
```
One possible solution to this is described in https://github.com/sparkletown/internal-sparkle-issues/issues/281

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,129 @@
+            and so we are defaulting it to LOCATION_INCREMENT_SECONDS
+            since that's how often this code runs
+          */
+          firestore
+            .collection("users")
+            .doc(userId)
+            .collection("visits")
+            .doc(locationName)
```
If you set just this part as a `const` that will give you the 'ref', you can then use it for both the `.update` and the `.set` without having to duplicate the 'path'.
```suggestion
          userLocationRef
```

mike-lvov replied:
fixed

0xdevalias reviewed (changes requested):
[CRO] Review based on what is in the diff here, I haven't gone through the comments in https://github.com/sparkletown/sparkle/pull/1108#pullrequestreview-580000514 1-by-1 to ensure they have all been addressed here.

mike-lvov commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,15 +21,40 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
+};
 
-export const RoomModal: React.FC<RoomModalProps> = ({ onHide, room }) => {
+export const RoomModalContent: React.FC<RoomModalContentProps> = ({
```
Yes

0xdevalias replied:
Cool cool, makes sense.

mike-lvov commented on src/components/molecules/EventDisplay/EventDisplay.tsx:
```diff
@@ -10,27 +11,32 @@ import { useRoom } from "hooks/useRoom";
 
 import "./EventDisplay.scss";
 
-interface EventDisplayProps {
+export type EventDisplayProps = {
```
Even though I disagree with this approach, I will change it since 2/3 team members agreed on this

mike-lvov replied:
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/components/molecules/EventDisplay/EventDisplay.tsx:
```diff
@@ -61,13 +69,15 @@ export const EventDisplay: React.FC<EventDisplayProps> = ({ event, venue }) => {
 
 type EnterRoomButtonProps = {
```
Even though I disagree with this approach, I will change it since 2/3 team members agreed on this

mike-lvov replied:
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,15 +21,40 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
 
 import "./RoomModal.scss";
 
-interface RoomModalProps {
+export type RoomModalProps = {
```
Even though I disagree with this approach, I will change it since 2/3 team members agreed on this

mike-lvov replied:
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/utils/room.ts:
```diff
@@ -0,0 +1,9 @@
+export type GetExternalRoomSlugProps = {
```
According to the article you were referring to, they recommend using types, when creating functions(why only functions?). Honestly, this article looks so opinionated and unreasonable, but as I understood we are going to follow their approach?

![image](https://user-images.githubusercontent.com/63194656/106889232-65246400-66f0-11eb-9df0-6e0c05f58bd5.png)

0xdevalias replied:
The case called out here is for props, which is an object, and falls into the ‘interface’ recommendations

0xdevalias replied:
fixed in 200c94f53638642cd3f70418f66b2accb8f392fe

mike-lvov commented on src/components/molecules/LiveSchedule/LiveSchedule.tsx:
```diff
@@ -42,9 +42,9 @@ const LiveSchedule: FC = () => {
-        key={`live-event-${index}`}
+    return events.map((event) => (
+      <EventDisplay
+        key={event.id ?? event.name}
```
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/components/molecules/VenueInfoEvents/VenueInfoEvents.tsx:
```diff
@@ -42,9 +42,9 @@ const VenueInfoEvents: React.FunctionComponent<PropsType> = ({
                 <div className="events-list events-list_monday">
                   {eventsNow &&
-                    eventsNow.map((event, idx) => (
+                    eventsNow.map((event) => (
```
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/components/organisms/SchedulePageModal/SchedulePageModal.tsx:
```diff
@@ -94,7 +94,7 @@ export const SchedulePageModal: FC<SchedulePageModalProps> = ({
     () =>
       orderedEvents[date]?.events.map((event) => (
```
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/components/organisms/SchedulePageModal/SchedulePageModal.tsx:
```diff
@@ -94,7 +94,7 @@ export const SchedulePageModal: FC<SchedulePageModalProps> = ({
       orderedEvents[date]?.events.map((event) => (
         <EventDisplay
-          key={event.id}
+          key={event.id ?? event.name}
```
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/components/molecules/VenueInfoEvents/VenueInfoEvents.tsx:
```diff
@@ -42,9 +42,9 @@ const VenueInfoEvents: React.FunctionComponent<PropsType> = ({
+                    eventsNow.map((event) => (
                       <EventDisplay
-                        key={event.name + idx}
+                        key={event.id ?? event.name}
```
changed in https://github.com/sparkletown/sparkle/pull/1129/commits/84027f4323263a9ef5e9e643914436010c1d2abe

mike-lvov commented on src/utils/room.ts:
```diff
@@ -0,0 +1,9 @@
+export const getExternalRoomSlug = ({
+  currentVenueName,
+  roomTitle,
+}: GetExternalRoomSlugProps) => `${currentVenueName}/${roomTitle}`;
```
roomSlug

0xdevalias replied:
fixed

mike-lvov commented on src/hooks/useRoom.ts:
```diff
@@ -2,38 +2,42 @@ import { useCallback, useMemo } from "react";
 import { useUser } from "hooks/useUser";
 
-export const useRoom = (room: Room) => {
+export const useRoom = (room: Room, currentVenueName: string) => {
```
fixed

mike-lvov commented on src/components/templates/PartyMap/components/RoomAttendance.tsx:
```diff
@@ -18,7 +18,7 @@ export const RoomAttendance: React.FC<RoomAttendanceProps> = ({
   room,
 }) => {
-  const { recentRoomUsers } = useRoom(room);
+  const { recentRoomUsers } = useRoom(room, venue.name);
```
fixed

mike-lvov commented on src/components/molecules/EventDisplay/EventDisplay.tsx:
```diff
@@ -48,8 +54,10 @@ export const EventDisplay: React.FC<EventDisplayProps> = ({ event, venue }) => {
+          {event.room && room && venue ? (
+            <EnterRoomButton room={room as Room} currentVenue={venue}>
+              {buttonText}
+            </EnterRoomButton>
```
fixed

mike-lvov commented on src/components/templates/Audience/Audience.tsx:
```diff
@@ -254,10 +255,11 @@ export const Audience: React.FunctionComponent = () => {
 
   return useMemo(() => {
     const takeSeat = (row: number | null, column: number | null) => {
```
added the comment

mike-lvov commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -37,91 +63,88 @@ export const RoomModal: React.FC<RoomModalProps> = ({ onHide, room }) => {
-              <h2 className="room-modal-title">{room.title}</h2>
-              <div className="room-modal-subtitle">{room.subtitle}</div>
-            </div>
+    <div className="container room-modal-container">
```
changed it here

0xdevalias replied:
Looks good. One more question around this in https://github.com/sparkletown/sparkle/pull/1129#discussion_r570716873 (but this comment seems resolved now, so i'll mark it as such)

mike-lvov commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,15 +21,40 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
+  const { enterRoom, recentRoomUsers } = useRoom(room, currentVenueName);
 
   useMousetrap({
     keys: "enter",
```
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570203533

mike-lvov commented on src/hooks/useKeyboardControls.ts:
```diff
@@ -10,7 +10,6 @@ export const MovementKeys = {
   down: ["down", "s"],
   left: ["left", "a"],
   right: ["right", "d"],
-  enter: ["enter"],
```
https://github.com/sparkletown/sparkle/pull/1129#discussion_r570203533

mike-lvov commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,15 +21,40 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
+  show,
+  currentVenue,
+}) => {
+  if (!room || !currentVenue) return null;
```
https://github.com/sparkletown/sparkle/pull/1129#discussion_r569927212

mike-lvov commented on src/components/templates/PartyMap/PartyMap.tsx:
```diff
@@ -30,6 +28,8 @@ export const PartyMap: React.FC = () => {
 
   const [selectedRoom, setSelectedRoom] = useState<Room | undefined>();
 
+  const hasSelectedRoom = !!selectedRoom;
```
https://github.com/sparkletown/sparkle/pull/1129#discussion_r569927212

mike-lvov commented:
Seems like I've addressed all of the comments. I reckon this was supposed to be a blocker for deploying master => production envs? If this is still the case, should only the main things get fixed and low-medium priorities skipped?

0xdevalias commented on src/utils/userLocation.ts:
```diff
@@ -1,81 +1,125 @@
+  roomSlug: string;
 };
 
-export const enterExternalRoom = ({ userId, room }: EnterExternalRoomProps) => {
-  trackLocationEntered({
+export const enterExternalRoom = ({
+  userId,
+  roomSlug,
+  roomUrl,
+}: EnterExternalRoomProps) => {
+  setLocationData({
     userId,
-    locationName: room.url,
+    locationName: roomSlug,
   });
```
Since `roomSlug` is just passed into `locationName` (which seems to be our standard way of referring to 'location slugs' in the code here), and also represents the 'location slug', it would probably be useful to change this to `locationName` as well.

0xdevalias replied:
fixed in 200c94f53638642cd3f70418f66b2accb8f392fe

0xdevalias reviewed (commented):
Thanks for fixing up the changes! Seems there may have been a couple of little bits missed.

The `interface/`type` comments are NB, though to keep consistency with the other changes made in this PR for the same 'pattern' (namely props objects used for the params in functions), it would be good to fix them up. I expect all of them would only take a minute or so tops, so shouldn't add much overhead.

0xdevalias commented on src/hooks/useRoom.ts:
```diff
@@ -2,38 +2,47 @@ import { useCallback, useMemo } from "react";
 import { useUser } from "hooks/useUser";
 
-export const useRoom = (room: Room) => {
+export type UseRoomProps = {
```
fixed in 200c94f53638642cd3f70418f66b2accb8f392fe

0xdevalias commented on src/components/templates/PartyMap/components/RoomModal/RoomModal.tsx:
```diff
@@ -20,108 +20,121 @@ import { RoomModalOngoingEvent, ScheduleItem } from "..";
+  show,
+  venue,
+}) => (
+  <Modal show={show} onHide={onHide}>
+    <div className="container room-modal-container">
+      {room && venue && <RoomModalContent room={room} venueName={venue.name} />}
+    </div>
+  </Modal>
```
It looks like we sort of keep the transition animation (so it seems it's applied to the container/similar), but because we're un-rendering the content within the modal, it looks like it quickly loses its height as part of the 'close modal' transition. We may want to give it a min-height, or do something else to 'solve this' in the future, but it doesn't seem like a big enough issue to block/delay merging this PR over. Would be good to backlog this for future fixing/similar.

mike-lvov replied:
https://github.com/sparkletown/internal-sparkle-issues/issues/306

0xdevalias reviewed (approved):
The only outstanding review comments seem to be:

- https://github.com/sparkletown/sparkle/pull/1129#discussion_r570728822
  - OOS/future cleanup
- https://github.com/sparkletown/sparkle/pull/1129#discussion_r570722366
  - @mike-lvov please add this explanation comment before merging
  
Based on that, i'm happy to approve this, with the caveat that the above comment is added before merging.

0xdevalias commented:
> Seems like I've addressed all of the comments. I reckon this was supposed to be a blocker for deploying master => production envs? If this is still the case, should only the main things get fixed and low-medium priorities skipped?

@mike-lvov I cleaned up a few of the little bits from my latest review quickly in https://github.com/sparkletown/sparkle/pull/1129/commits/200c94f53638642cd3f70418f66b2accb8f392fe

Based on that, the only outstanding thing needed before merging is adding the comment mentioned in https://github.com/sparkletown/sparkle/pull/1129#pullrequestreview-584033564
